### PR TITLE
Implement update + read flow for model version support in Detector

### PR DESCRIPTION
**Pull request by srrokib**
### Notes (Rev 2)
- Add more tests to `test_validation_helpers.py` and `test_update_worker_helpers.py`, in order to fully stub out `get_model_version` responses
- Rename `validate_model_versions_for_detector_create` -> `validate_model_versions_for_detector_model`
- Avoid KeyError for `model_version["arn"]`
- Fix comment in `update_worker_helpers.py`
- Remove duplicate get_external_models instance
- Diff: https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-frauddetector/pull/29/commits/179f68c444ed637dce428fdd17c0b258f8883b93

### Notes
- Add  `validate_model_versions_for_detector_create` to update handler
- Add model versions into output model for read handler
- Update unit test data

#### Unit Tests (updated)

`./run_unit_tests`

```
...
...
Required test coverage of 89.0% reached. Total coverage: 89.31%

========================================================================================================= 213 passed in 31.78s =========================================================================================================
```

#### Contract Tests

`./cfn_server detector` and `cfn test`

```
...
...

============================================================================================== 14 passed, 2 skipped in 151.42s (0:02:31) ===============================================================================================
```

### Pre-commit

`pre-commit run --all-files `

```
black....................................................................Passed
flake8...................................................................Passed
Check for case conflicts.................................................Passed
Detect Private Key.......................................................Passed
Fix End of Files.........................................................Passed
Mixed line ending........................................................Passed
Trim Trailing Whitespace.................................................Passed
Pretty format JSON.......................................................Passed
Check for merge conflicts................................................Passed
Check Yaml...............................................................Passed
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/model_helpers.py [hunk 1]
```diff
@@ -105,6 +104,11 @@ def get_model_for_detector(frauddetector_client, detector, model: models.Resourc
             # we should never see this block get executed
             raise exceptions.NotFound("associatedModel", model_endpoint)
         associated_models.append(models.Model(Arn=external_models[0].get("arn", "not/found")))
+
+    model_versions: List[dict] = desired_detector_version.get("modelVersions", [])
+    for model_version in model_versions:
+        associated_models.append(models.Model(Arn=model_version["arn"]))
```
Comment (srrokib) [inline aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/model_helpers.py:110]: Sanity check: this will fetch current state of events for model version(s) attached to Detector Version. This will cover out of band updates to this Detector?

Comment (privetjr) [reply aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/model_helpers.py:110]: If the customer attaches a different model version to this detector outside of CFN, will the next read from CFN will have that new model version's arn as an AssociatedModel, yes.

One comment here, if for some reason, AFD does not return `arn` for the model version, we would get KeyError on line 110. Thoughts on avoiding a KeyError for `Arn=model_version["arn"]`?

File: aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ def validate_dependencies_for_detector_update(
     afd_client, model: models.ResourceModel, previous_model: models.ResourceModel
 ):
     # TODO: revisit  this validation when/if we support in-place teardown
-    #   For now, throw bad request for unsupported event type update and validate external models
+    #   For now, throw bad request for unsupported event type update and external models, and model versions
```
Comment (privetjr) [inline aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py:23]: nit: Without the `validate` word, this seems worded oddly, since we're doing slightly different things with event type checks vs. model checks.

```
For now, throw bad request for unsupported event type update, and validate external models + model versions
```

File: aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py [hunk 3]
```diff
@@ -35,6 +35,7 @@ def validate_dependencies_for_detector_update(
         if not get_event_types_succeeded:
             raise exceptions.NotFound("detector.EventType", event_type_name)
     validation_helpers.validate_external_models_for_detector_model(afd_client, model)
+    validation_helpers.validate_model_versions_for_detector_create(afd_client, model)
```
Comment (privetjr) [inline aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py:38]: nit: just realized we did not rename this method. Since we use this for `detector_update` as well as `detector_create`, should we rename to `validate_model_versions_for_detector_model` ?

File: aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_model_helpers.py [hunk 4]
```diff
@@ -74,11 +74,15 @@ def test_get_model_for_detector():
     mock_afd_client.get_rules = MagicMock(return_value=get_rules_response)
     mock_afd_client.get_outcomes = MagicMock(return_value=get_outcomes_response)
     mock_afd_client.get_external_models = MagicMock(return_value=get_external_models_response)
+    mock_afd_client.get_external_models = MagicMock(return_value=get_external_models_response)
```
Comment (privetjr) [inline aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_model_helpers.py:77]: Don't need this duplicate line (and don't need to mock get_model_versions, since model version arn is returned in get_detector_version)

File: aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py (hunk 3 above)
Comment (srrokib) [reply aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py:38]: Agreed. Revising

**Review by privetjr — CHANGES_REQUESTED** [review]
Couple of comments inline, and one extra callout. 

I think we're missing some unit tests details for both create_worker_helpers and update_worker_helpers to confirm that we are validating model versions (by calling get_model_version)

e.g. lines like this: https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-frauddetector/blob/master/aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_create_worker_helpers.py#L48

File: aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_model_helpers.py (hunk 4 above)
Comment (srrokib) [reply aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_model_helpers.py:77]: Good catch. Removing, I was initially going after an API call but then it made sense to rely on the response from `GetDetectorVersion`

**Comment by srrokib** [comment]
@privetjr 

>Couple of comments inline, and one extra callout.
>
>I think we're missing some unit tests details for both create_worker_helpers and update_worker_helpers to confirm that we are validating model versions (by calling get_model_version)
>
>e.g. lines like this: https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-frauddetector/blob/master/aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_create_worker_helpers.py#L48

- `create_worker_helpers` test has been included for detector create with 1 MV, `test_validate_dependencies_for_detector_create_happy_case_with_model_version ` https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-frauddetector/blob/59c185e30f498601141198fcba8abd02a35c3d12/aws-frauddetector-detector/src/aws_frauddetector_detector/tests/helpers/test_create_worker_helpers.py#L183

- `update_worker_helpers` is calling `validate_model_versions_for_detector_create`, which has all scenarios covered in unit_tests.py, happy case + exceptions. I do see value in adding a happy case test in `test_update_worker_helpers.py` that verifies an invocation of `get_model_version` in the update flow–this is currently missing.

File: aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py (hunk 2 above)
Comment (srrokib) [reply aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/update_worker_helpers.py:23]: Good catch. Revising

**Comment by privetjr** [comment]
> TLDR: revising `test_update_worker_helpers.py` to add a happy case test

We might not need a new test, you can probably just mock a DV with an MV and stick the assert on one of the existing tests. Just as long as we have a test that confirms we check for MV when there is an AssociatedModel that is an MV.

File: aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/model_helpers.py (hunk 1 above)
Comment (srrokib) [reply aws-frauddetector-detector/src/aws_frauddetector_detector/helpers/model_helpers.py:110]: We could definitely get a `KeyError` here, but that would be a highly unexpected scenario given that we expect an array of ModelVersions from `get_detector_version`, and if that array is non empty, it should contain an ARN.

It does not hurt to add a check here for key error, and add some context that we do not expect that line to execute in the comments.

**Review by privetjr — APPROVED** [review]
LGTM, thank you for addressing comments.

**Review by sahilper — APPROVED** [review]
